## Calculate the weight of individual element to synthesize bulk ingots of alloys

This notebook helps in determining the weights of elements, in grams, to preparte bulk ingots of alloys. If we want to synthesize Co$_2$FeGe, for example, then we need the weights of Co, Fe, and Ge. The general idea is to find the total atomic weight of the alloy and use it to determine the weight fraction of each elements. Then the weight fraction can be used to find the weight of the elements (in grams) to make an alloy of desired weight (for example 5 grams).

***Getting data:*** We use the periodic table from Github page uploaded by Goodman Sciences. We direclty import the data and read this as pandas dataframe.

In [1]:
path = "https://gist.github.com/GoodmanSciences/c2dd862cd38f21b0ad36b8f96b4bf1ee/raw/1d92663004489a5b6926e944c1b3d9ec5c40900e/Periodic%2520Table%2520of%2520Elements.csv"
import pandas as pd
df = pd.read_csv(path)
df.head()

AtomicNumber    Element Symbol  AtomicMass  NumberofNeutrons  \
0             1   Hydrogen      H       1.007                 0   
1             2     Helium     He       4.002                 2   
2             3    Lithium     Li       6.941                 4   
3             4  Beryllium     Be       9.012                 5   
4             5      Boron      B      10.811                 6   

   NumberofProtons  NumberofElectrons  Period  Group  Phase  ...  \
0                1                  1       1    1.0    gas  ...   
1                2                  2       1   18.0    gas  ...   
2                3                  3       2    1.0  solid  ...   
3                4                  4       2    2.0  solid  ...   
4                5                  5       2   13.0  solid  ...   

  FirstIonization   Density MeltingPoint BoilingPoint NumberOfIsotopes  \
0         13.5984  0.000090       14.175        20.28              3.0   
1         24.5874  0.000179          NaN         4.22              5.0   
2          5.3917  0.534000      453.850      1615.00              5.0   
3          9.3227  1.850000     1560.150      2742.00              6.0   
4          8.2980  2.340000     2573.150      4200.00              6.0   

   Discoverer    Year  SpecificHeat  NumberofShells  NumberofValence  
0   Cavendish  1766.0        14.304               1              1.0  
1     Janssen  1868.0         5.193               1              NaN  
2   Arfvedson  1817.0         3.582               2              1.0  
3  Vaulquelin  1798.0         1.825               2              2.0  
4  Gay-Lussac  1808.0         1.026               2              3.0  

[5 rows x 28 columns]

When entering the symbols of elements, it is a good idea to make it case-insensitive. So, we create a column, where we store the elements symbol in all uppercase.

In [2]:
df["Symbol_upper"] = df["Symbol"].str.upper()


The cell below takes the elements name and their respective atomic fractions as keyboard inputs. There is no limit on the number of elements, but to break the process, it is required to enter 0. An example of Co$_2$FeGe is shown.

In [11]:
dic1 = {}       #create a dictionary with desired elements
tot_weight = 0
while True:
      element_ = input("Enter the name of element, enter 0 to break:").upper()
      if str(element_) == "0":
          break
      fraction_ = float(input("Enter the atomic fraction of the element:"))
      dic1[element_] = df.loc[df["Symbol_upper"]==element_,"AtomicMass"].item()*fraction_   # atomic wt. of element scaled by their atomic fraction in the alloy
      tot_weight = tot_weight + (dic1[element_]) # calculate total weight for the alloy

Enter the name of element, enter 0 to break:co
Enter the atomic fraction of the element:2
Enter the name of element, enter 0 to break:fe
Enter the atomic fraction of the element:1
Enter the name of element, enter 0 to break:ge
Enter the atomic fraction of the element:1
Enter the name of element, enter 0 to break:0


The cell below takes the desired sample weight as keyboard input and calculates weight %, atomic %, weight in grams and stores the result into a dictionary.

If different sample weight is needed, one can just run the cells below. No need to run above cells, given that they have run already.

In [28]:
samplewt = float(input("Enter the sample weight you want in grams: "))  
print("The desired sample weight is entered as {} grams".format(samplewt))
dic_result = {}
for key in dic1.keys(): 
    wt_per = round((dic1[key]/tot_weight)*100,3)
    at_per = round(wt_per/df[df["Symbol_upper"]==key]["AtomicMass"].iloc[0],3)
    element_wt = round(wt_per*samplewt/100,3)
    dic_result[df[df["Symbol_upper"]==key]["Symbol"].iloc[0]] = [wt_per, at_per, element_wt]
    

Enter the sample weight you want in grams: 5
The desired sample weight is entered as 5.0 grams


The dictionary is then converted into pandas dataframe, which helps in visualizing the result.

In [37]:
result_df = pd.DataFrame(dic_result, index=['Weight %','Atomic %','Desired Weight(in grams)']).transpose()
result_df['Atomic %'] = (result_df['Atomic %']/result_df['Atomic %'].sum())*100
result_df.index.name = "Element"
result_df.loc['Total'] = result_df.sum( numeric_only=True,axis=0)
print("The desired results for {} grams sample".format(samplewt))
result_df

The desired results for 5.0 grams sample


Weight %  Atomic %  Desired Weight(in grams)
Element                                              
Co         47.845      50.0                     2.392
Fe         22.669      25.0                     1.133
Ge         29.486      25.0                     1.474
Total     100.000     100.0                     4.999

Hence, to make 5 grams of Co$_2$FeGe, one requires 2.392 grams Co, 1.133 grams Fe, and 1.474 grams Ge.